packages/flutter_tools/lib/src/daemon.dart

```diff
@@ -0,0 +1,242 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'base/common.dart';
+import 'base/io.dart';
+import 'base/logger.dart';
+import 'base/utils.dart';
+import 'convert.dart';
+import 'globals.dart' as globals;
+
+/// Parse binary streams in the JSON RPC format understood by the daemon, and
+/// convert it into a stream of JSON RPC messages.
+Stream<Map<String, Object?>> _convertInputStream(Stream<List<int>> inputStream) {
+  return utf8.decoder.bind(inputStream)
+  .transform<String>(const LineSplitter())
+  .where((String line) => line.startsWith('[{') && line.endsWith('}]'))
+  .map<Map<String, Object?>?>((String line) {
+    line = line.substring(1, line.length - 1);
+    return castStringKeyedMap(json.decode(line));
+  })
+  .where((Map<String, Object?>? entry) => entry != null)
+  .cast<Map<String, Object?>>();
+}
+
+/// A stream that a [DaemonConnection] uses to communicate with each other.
+abstract class DaemonStreams {
+  /// Stream that contains input to the [DaemonConnection].
+  Stream<Map<String, Object?>> get inputStream;
+
+  /// Outputs a message through the connection.
+  void send(Map<String, Object?> message);
+
+  /// Cleans up any resources used.
+  Future<void> dispose() async { }
+}
+
+/// A [DaemonStream] that uses stdin and stdout as the underlying streams.
+class StdioDaemonStreams extends DaemonStreams {
+  StdioDaemonStreams([Stdio? stdio]) :
+    _stdio = stdio ?? globals.stdio,
+    inputStream = _convertInputStream((stdio ?? globals.stdio).stdin);
+
+  final Stdio _stdio;
+
+  @override
+  final Stream<Map<String, Object?>> inputStream;
+
+  @override
+  void send(Map<String, Object?> message) {
+    _stdio.stdoutWrite(
+      '[${json.encode(message)}]\n',
+      fallback: (String message, Object? error, StackTrace stack) {
+        throwToolExit('Failed to write daemon command response to stdout: $error');
+      },
+    );
+  }
+}
+
+/// A [DaemonStream] that uses [Socket] as the underlying stream.
+class TcpDaemonStreams extends DaemonStreams {
+  /// Creates a [DaemonStreams] with an existing [Socket].
+  TcpDaemonStreams(
+    Socket socket, {
+    Logger? logger,
+  }): _logger = logger ?? globals.logger {
+    _socket = Future<Socket>.value(_initializeSocket(socket));
+  }
+
+  /// Connects to a remote host and creates a [DaemonStreams] from the socket.
+  TcpDaemonStreams.connect(
```

Comment:
Is this used?

Reply:
This will be used in my next PR, would you want me to remove it from here, and add the change in my next PR instead?

Reply:
No, you can leave it in this PR, wanted to make sure it wasn't leftover from something else you tried.